components/provisioning/org.wso2.carbon.identity.provisioning/src/main/java/org/wso2/carbon/identity/provisioning/OutboundProvisioningManager.java

```diff
@@ -824,12 +826,22 @@ protected boolean canUserBeProvisioned(ProvisioningEntity provisioningEntity,
             return true;
         }
 
-        String userName = getUserName(provisioningEntity.getAttributes());
-        List<String> roleListOfUser = getUserRoles(userName, tenantDomain);
+        if(provisioningEntity.getAttributes() != null) {
+            String userName = getUserName(provisioningEntity.getAttributes());
+            List<String> roleListOfUser = getUserRoles(userName, tenantDomain);
+
+            List<String> newRoleListOfUser = provisioningEntity.getAttributes().get(ClaimMapping.build
+                    (IdentityProvisioningConstants.GROUP_CLAIM_URI, null, null, false));
 
-        for (String provisionByRole : provisionByRoleList) {
-            if (roleListOfUser.contains(provisionByRole)) {
-                return true;
+            if(newRoleListOfUser != null) {
+                 roleListOfUser = Stream.concat(roleListOfUser.stream(), newRoleListOfUser.stream()).distinct()
+                         .collect(Collectors.toList());
+            }
+
+            for (String provisionByRole : provisionByRoleList) {
+                if (roleListOfUser.contains(provisionByRole)) {
```

Comment:
We are just checking a particular role is available in two another list?
If so, it is not good to create another list with distinct strings. This creates large list, and add garbage pressure.
Best is to check on original two lists one after another, which does not create new list.